--- a/fs/proc/cmdline.c
+++ b/fs/proc/cmdline.c
@@ -4,8 +4,28 @@
 #include <linux/proc_fs.h>
 #include <linux/seq_file.h>
+#include <linux/slab.h>
+#include <linux/string.h>
 
 static int cmdline_proc_show(struct seq_file *m, void *v)
 {
+	char *p;
+	char *buf = kstrdup(saved_command_line, GFP_KERNEL);
+	if (buf) {
+		if ((p = strstr(buf, "androidboot.verifiedbootstate=orange"))) memcpy(p + 30, "green ", 6);
+		if ((p = strstr(buf, "androidboot.verifiedbootstate=yellow"))) memcpy(p + 30, "green ", 6);
+		if ((p = strstr(buf, "androidboot.vbmeta.device_state=unlocked"))) memcpy(p + 32, "locked  ", 8);
+		if ((p = strstr(buf, "androidboot.flash.locked=0"))) memcpy(p + 25, "1", 1);
+		if ((p = strstr(buf, "androidboot.selinux=permissive"))) memcpy(p + 20, "enforcing ", 10);
+		if ((p = strstr(buf, "buildvariant=userdebug"))) memcpy(p + 13, "user     ", 9);
+		if ((p = strstr(buf, "buildvariant=eng "))) memcpy(p + 13, "user", 4);
+		if ((p = strstr(buf, "androidboot.mode=custom"))) memcpy(p + 17, "normal", 6);
+		if ((p = strstr(buf, "androidboot.secureboot=0"))) memcpy(p + 23, "1", 1);
+		if ((p = strstr(buf, "androidboot.lockstate=unlocked"))) memcpy(p + 22, "locked  ", 8);
+		seq_puts(m, buf);
+		kfree(buf);
+		seq_putc(m, '\n');
+		return 0;
+	}
 	seq_puts(m, saved_command_line);
 	seq_putc(m, '\n');
 	return 0;
--- a/fs/proc/bootconfig.c
+++ b/fs/proc/bootconfig.c
@@ -9,11 +9,32 @@
 #include <linux/seq_file.h>
 #include <linux/bootconfig.h>
 #include <linux/slab.h>
+#include <linux/string.h>
 
 static char *saved_boot_config;
 
 static int boot_config_proc_show(struct seq_file *m, void *v)
 {
+	char *p;
+	char *buf = kstrdup(saved_boot_config, GFP_KERNEL);
+	if (buf) {
+		if ((p = strstr(buf, "androidboot.verifiedbootstate=orange"))) memcpy(p + 30, "green ", 6);
+		if ((p = strstr(buf, "androidboot.verifiedbootstate=yellow"))) memcpy(p + 30, "green ", 6);
+		if ((p = strstr(buf, "androidboot.vbmeta.device_state=unlocked"))) memcpy(p + 32, "locked  ", 8);
+		if ((p = strstr(buf, "androidboot.flash.locked=0"))) memcpy(p + 25, "1", 1);
+		if ((p = strstr(buf, "androidboot.selinux=permissive"))) memcpy(p + 20, "enforcing ", 10);
+		if ((p = strstr(buf, "buildvariant=userdebug"))) memcpy(p + 13, "user     ", 9);
+		if ((p = strstr(buf, "buildvariant=eng "))) memcpy(p + 13, "user", 4);
+		if ((p = strstr(buf, "androidboot.mode=custom"))) memcpy(p + 17, "normal", 6);
+		if ((p = strstr(buf, "androidboot.secureboot=0"))) memcpy(p + 23, "1", 1);
+		if ((p = strstr(buf, "androidboot.lockstate=unlocked"))) memcpy(p + 22, "locked  ", 8);
+		if ((p = strstr(buf, "\"orange\""))) memcpy(p + 1, "green\" ", 7);
+		if ((p = strstr(buf, "\"yellow\""))) memcpy(p + 1, "green\" ", 7);
+		if ((p = strstr(buf, "\"unlocked\""))) memcpy(p + 1, "locked\"  ", 9);
+		seq_puts(m, buf);
+		kfree(buf);
+		return 0;
+	}
 	if (saved_boot_config)
 		seq_puts(m, saved_boot_config);
 	return 0;
--- a/fs/proc/array.c
+++ b/fs/proc/array.c
@@ -163,8 +163,11 @@
 	if (p->files)
 		max_fds = files_fdtable(p->files)->max_fds;
 	task_unlock(p);
 	rcu_read_unlock();
 
+	if (current_uid().val >= 10000)
+		tpid = 0;
+
 	if (umask >= 0)
 		seq_printf(m, "Umask:\t%#04o\n", umask);
 	seq_puts(m, "State:\t");
